// aifactory/aifactory.ai.gen.util-1.0.0.js
/*
	AI factory; GPT / LLM / CLAUDE 3
	See README.md

    Example 1: Conversation Assistant
    Check conversations that part of and check for messages that have 
    [@genai:{user message}]
    Post a comment.

    In event mind to a particular conversation.guid for now.

	References:
	- https://docs.anthropic.com/claude/reference/messages_post
	- https://docs.anthropic.com/claude/docs/models-overview

	
	curl https://api.anthropic.com/v1/messages \
		--header "x-api-key: $ANTHROPIC_API_KEY" \
		--header "anthropic-version: 2023-06-01" \
		--header "content-type: application/json" \
		--data \
	'{
		"model": "claude-3-opus-20240229",
		"max_tokens": 1024,
		"messages": [
			{"role": "user", "content": "Hello, world"}
		]
	}'
*/

var entityos = require('entityos')
var _ = require('lodash')
var moment = require('moment');

module.exports = 
{
	VERSION: '1.0.0',

	init: function (param)
	{
		entityos.add(
		{
			name: 'ai-gen-util-get-settings',
			code: function (param)
			{
				let data = entityos.get({scope: '_data'});
				const settings = entityos.get({scope: '_settings'});

				let services = settings.ai.services;

				let service;
				let model;
			
				if (data.service != undefined)
				{
					service = _.find(services, function (service)
					{
						return (service.name == data.service)
					});
				}

				if (service == undefined && data.model != undefined) 
				{
					_.each(services, function(_service)
					{
						if (model == undefined)
						{
							model = _.find(_service.models, function (model)
							{
								return (model.name == data.model)
							});

							if (model != undefined)
							{
								service = _service
							}
						}
					});
				}

				if (service == undefined)
				{
					service = _.find(services, function (service)
					{
						return (service.default)
					});
				}

				if (service != undefined && model == undefined) 
				{
					model = _.find(service.models, function (model)
					{
						return (model.name == data.model)
					});
				}

				if (model == undefined)
				{
					model = _.find(service.models, function (model)
					{
						return (model.default)
					});
				}

				if (service != undefined)
				{
					if (service.namespaces != undefined)
					{
						_.each(service.namespaces, function (namespace)
						{
							var aifactory = require('aifactory/aifactory.' + namespace + '.js');

							if (_.has(aifactory, 'init'))
							{
								aifactory.init();
							}
						})
					}
				}

				return {model: model, service: service};
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-chat',
			code: function (param)
			{          
                const settings = entityos.get({scope: '_settings'});

				let aiSettings = _.get(param, 'settings');

				if (aiSettings == undefined)
				{
					//use ai-gen-util-get-model with modelName
				}
				else
				{			
					let messages = _.get(param, 'messages', {});

					if (_.get(messages, 'system') == undefined)
					{
						messages.system = _.get(settings, 'ai.defaults.messages.system');
					}

					const host = _.get(aiSettings, 'service.host'); 

					if (host == undefined)
					{
						entityos.invoke('util-end', {error: 'No host settings.'});
					}
					else
					{
						if (messages.user == undefined)
						{
							entityos.invoke('util-end', {error: 'No user message (prompt text)'});
						}
						else
						{
							let hostName = host.name;
							let hostPath = host.path;

							let keyPath = _.get(aiSettings, 'service.keypath');
							
							let apiKey = _.get(settings, keyPath);

							if (hostName == undefined || hostPath == undefined || apiKey == undefined)
							{
								entityos.invoke('util-end', {error: 'url, urlpath, or apikey is missing.'});
							}
							else
							{
								// Need to make headers, asSettings/Service driven
								let headers =
								{
									'x-api-key': apiKey,
									'Content-Type': 'application/json',
									Accept: 'application/json'
								}

								if (_.get(aiSettings, 'service.headers') != undefined)
								{
									headers = _.assign(headers, aiSettings.service.headers)
								}

								const maxTokensDefault = _.get(settings, 'ai.defaults.maxtokens', 1000);
								const temperatureDefault = _.get(settings, 'ai.defaults.temperature', 0.7);

								const maxTokensDefaultService = _.get(aiSettings, 'service.defaults.maxtokens', maxTokensDefault);
								const tempatureDefaultService = _.get(aiSettings, 'service.defaults.temperature', temperatureDefault);
									
								const maxTokens = _.get(param, 'maxTokens', maxTokensDefaultService);
								const temperature = _.get(param, 'temperature', tempatureDefaultService);

								var sendOptions = 
								{
									hostname: hostName,
									path: hostPath,
									headers: headers,
									action: 'POST',
									callback: 'ai-gen-util-chat-response',
									onComplete: param.onComplete,
									data:
									{
										model: aiSettings.model.name
									}
								}

								sendOptions.data.system =  param.messages.system;
								sendOptions.data.max_tokens = maxTokens;
								sendOptions.datatemperature = temperature;

								sendOptions.data.messages =
								[
									{ role: 'user', content: messages.user },
									{ role: 'system', content: messages.system }
								]

								entityos._util.send(sendOptions, 'ai-gen-util-chat-response');
							}
						}
					}
				}
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-chat-response',
			code: function (param, response)
			{	
				//console.log(response.data);
				const completion = response.data;
				
				if (completion.type == 'error')
				{
					param.error = response.error;
					entityos.invoke('util-end', {error: completion.error});
				}
				else
				{
					_.set(param, 'messages.response', completion.content[0].text);

					if (_.get(param, 'onComplete') != undefined)
					{
						entityos._util.onComplete(param)
					}
					else
					{
						entityos.invoke('util-end', param);
					}
				}
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-file-to-base64',
			code: function (param)
			{	
				var event = entityos.get({scope: '_event'});

				const fs = require('fs');
				const path = require('path');

				const base64Image = fs.readFileSync(event.filename, { encoding: 'base64' });
				const baseFilename = path.basename(event.filename, path.extname(event.filename));
				const outputPath = path.join(path.dirname(event.filename), baseFilename + '-base64.txt');
				fs.writeFileSync(outputPath, base64Image);
				entityos.invoke('util-end', {message: outputPath});
			}
		});	

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download',
			code: function (param)
			{	
				var event = entityos.get({scope: '_event'});
				var attachmentUUID = _.get(event, 'attachment.uuid');
				var attachmentIndex = _.get(event, 'attachment.index', 0);

				let filters = [];

				let sorts = [];

				if (attachmentUUID != undefined)
				{
					filters.push(
					{
						field: 'guid',
						comparison: 'EQUAL_TO',
						value: attachmentUUID
					});

					attachmentIndex = 0;
				}
				else
				{
					filters = _.get(event, 'attachment.filters', [
					{
						field: 'filename',
						comparison: 'TEXT_IS_LIKE',
						value: '.pdf'
					}]);

					sorts.push(
					{
						name: 'createddate',
						direction: 'desc'
					});
				}

				const rows = (attachmentIndex + 1)

				entityos.cloud.search(
				{
					object: 'core_attachment',
					fields:
					[
						{name: 'filename'},
						{name: 'guid'},
						{name: 'object'},
						{name: 'objectcontext'}
					],
					filters: filters,
					sorts: sorts,
					rows: rows,
					callback: 'ai-gen-util-attachment-download-response'
				});
			}
		});		

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download-response',
			code: function (param, response)
			{
				var event = entityos.get({scope: '_event'});
				var attachmentUUID = _.get(event, 'attachment.uuid');
				var attachmentIndex = _.get(event, 'attachment.index', 0);

				if (response.status == 'ER')
				{
					entityos.invoke('util-end', {error: 'Error retrieving the attachment [' + attachmentUUID + ']'}, '401');
				}
				else
				{
					if (response.data.rows.length == 0)
					{
						entityos.invoke('util-end', {error: 'Bad attachment.uuid [' + attachmentUUID + ']'}, '401');
					}
					else
					{
						event.attachment = response.data.rows[attachmentIndex];
						entityos.set({scope: '_event', value: event});

						entityos._util.attachment.download({
							id: event.attachment.id,
							object: event.attachment.object,
							onComplete: 'ai-gen-util-attachment-download-process'
						});
					}
				}
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download-process',
			code: function (param)
			{
				var event = entityos.get({scope: '_event'});
				var attachment = _.get(event, 'attachment');
				//_.set(event, 'file.base64', _.truncate(param.base64, 50));
				_.set(event, 'file.base64', param.base64);
				_.set(event, 'file.name',
						'eos-' + attachment.object + '-' + attachment.objectcontext + '-' + attachment.id + '_' + attachment.filename
				);

				entityos.set({scope: '_event', value: event});
	
				if (event.controller == 'ai-gen-util-attachment-download-file-upload')
				{
					entityos.invoke('ai-gen-util-attachment-download-file-upload-process', param);
				}
				else
				{
					if (_.get(param, 'onComplete') != undefined)
					{
						entityos._util.onComplete(param)
					}
					else
					{
						entityos.invoke('util-end', event);
					}
				}
			}
		});

		

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download-file-upload',
			code: function (param)
			{
				entityos.invoke('ai-gen-util-attachment-download', param);
				// Based on the event.controller = ai-gen-util-attachment-download-file-upload'
				// - it invokes: 'ai-gen-util-attachment-download-file-upload-process'
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download-file-upload-process',
			code: function (param)
			{	
				_.set(param, 'onComplete', 'ai-gen-util-attachment-download-file-upload-vector-store-link');
				entityos.invoke('ai-gen-util-file-upload', param);
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download-file-upload-vector-store-link',
			code: function (param)
			{			
				_.set(param, 'onComplete', 'ai-gen-util-attachment-download-file-upload-finalise');
				entityos.invoke('ai-gen-util-vector-store-link-file', param)
			}
		});

		entityos.add(
		{
			name: 'ai-gen-util-attachment-download-file-upload-finalise',
			code: function (param)
			{
				var event = entityos.get({scope: '_event'});
				let data = {
					file:
					{
						id: _.get(event, 'file.id'),
						name: _.get(event, 'file.name')
					}
				}

				entityos.invoke('util-end', data);
			}
		});	
	}
}
